### PR TITLE
Allow pausing

### LINKED ISSUE #47: Allow pausing game
# Feature Request

Allow pausing the game (i.e., saving the current state of the game somewhere) and quitting. The next time `nudoku` starts, either the paused game can be resumed, or a new game can be played.

### PR DESCRIPTION
**Pull request by dogukanteber**
Resolves #47. Now the users can save the game and quit. If the game is saved, the next time they open nudoku, a pop-up asks them to resume where they left off or create a new game. Currently, there is no coloring in the pop-up. If there is a necessity to add colors, it is not a difficult task. It can be added easily as long as which colors is going to be used are provided.

### CODE REVIEW

**Comment by jubalh** [comment]
Thanks! I'll try to review it soon!

File: src/main.c [hunk 1]
```diff
@@ -71,7 +73,17 @@ static char* g_outputFilename = NULL;		/* in case -p/-i flag we get a filename p
 static int   g_sudokuCount = 1;				/* in case of -n we can the numbers of sudoku that should end up in the PDf (-p) */
 static bool  g_outIsPDF;
 static DIFFICULTY g_level = D_EASY;
-static WINDOW *grid, *infobox, *status;
+static WINDOW *grid, *infobox, *status, *resume;
+static bool is_board_solved = false;		/* if the board is solved, we do not save the state */
```
Comment (jubalh) [inline src/main.c:77]: I would propose to remove this variable and not check if the user solved/not ask him if he wants to save.
I would like to just have an option "lets save". And the user can do this when he wants but don't add the overhead of tracking this and asking him.

File: src/main.c [hunk 2]
```diff
@@ -333,6 +345,8 @@ static void init_windows(void)
 	wprintw(infobox, _(" r - Redraw\n"));
 	wprintw(infobox, _(" S - Solve puzzle\n"));
 	wprintw(infobox, _(" x - Delete number\n"));
+	wprintw(infobox, _(" g - Save\n"));	
+	wprintw(infobox, _(" G - Save and quit\n"));
```
Comment (jubalh) [inline src/main.c:349]: I think `save` is enough. The user can use `save` and then `quit`.

File: src/main.c [hunk 3]
```diff
@@ -400,6 +414,145 @@ static void fill_grid(char *user_board, char *plain_board, int x_cursor, int y_c
 	}
 }
 
+char* get_paused_file_path(void) {
+	const char* home_path = getenv("HOME");
+	size_t len_home_path = strlen(home_path);
+
+	const char* dir_name = "/.nudoku/";
```
Comment (jubalh) [inline src/main.c:421]: I think we should use something like: https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html
Most likely: ` $XDG_STATE_HOME defines the base directory relative to which user-specific state files should be stored. If $XDG_STATE_HOME is either not set or empty, a default equal to $HOME/.local/state should be used. `

File: src/main.c [hunk 4]
```diff
@@ -400,6 +414,145 @@ static void fill_grid(char *user_board, char *plain_board, int x_cursor, int y_c
...
+
+	const char* file_name = "game_state.txt";
+	size_t len_file = strlen(file_name);
+
+	char* file_path = malloc(len_home_path + len_dir_name + len_file + 1);
+	strcpy(file_path, dir_path);
+	strcat(file_path, file_name);
+	
+	free(dir_path);
+	return file_path;
```
Comment (jubalh) [inline src/main.c:441]: When is this file removed again? I think in `is_game_saved` you test for the presence of it. But I don't see where it get's removed when the game is finished.

**Review by jubalh — CHANGES_REQUESTED** [review]
The code looks good! Thank you!
I made some comments about where I think there need to be changes.
Sorry for the long delay! :(

**Comment by jubalh** [comment]
@dogukanteber ping

**Comment by dogukanteber** [comment]
Wow! It has been quite some time. I completely forgot about this PR @jubalh. My apologies. I will try to allocate some time to finish up this PR.

**Comment by merv1n34k** [comment]
@dogukanteber @jubalh Do you mind if I try to finish up this PR? *(will create a new one though)*
I have more free time than usual for a few days, and with the help of @dogukanteber's code it looks doable for me.

**Comment by jubalh** [comment]
Sure! Go for it :)

**Pull request closed without merging**